File: src/main/scala/codechicken/multipart/ItemMultiPart.scala
package codechicken.multipart

import codechicken.lib.vec.{Rotation, Vector3}
import codechicken.multipart.api.part.TMultiPart
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.{Item, ItemStack, ItemUseContext}
import net.minecraft.util._
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World

/**
 * Java class implementation
 */
abstract class JItemMultiPart(properties: Item.Properties) extends Item(properties) with TItemMultiPart

/**
 * Simple multipart item class for easy placement. Simply override the newPart function and it the part will be added to the block space if it passes the occlusion tests.
 */
trait TItemMultiPart extends Item {
    def getHitDepth(vhit: Vector3, side: Int): Double =
        vhit.copy.scalarProject(Rotation.axes(side)) + (side % 2 ^ 1)

    override def onItemUse(context: ItemUseContext): ActionResultType = {
        val stack = context.getPlayer.getHeldItem(context.getHand)
        val world = context.getWorld
        val player = context.getPlayer
        var pos = context.getPos
        val side = context.getFace.getIndex
        val vhit = new Vector3(context.getHitVec).subtract(pos)
        val d = getHitDepth(vhit, side)

        def place(offset: Boolean): ActionResultType = {
            val part = newPart(stack, player, world, pos, side, vhit)
            if (part == null || !TileMultipart.canPlacePart(context, part, offset)) return ActionResultType.FAIL

            if (!world.isRemote) {
                TileMultipart.addPart(world, pos, part)
                val sound = part.getPlacementSound(stack, player)
                if (sound != null) {
                    world.playSound(null, pos, sound.getPlaceSound,
                        SoundCategory.BLOCKS, (sound.getVolume + 1.0F) / 2.0F, sound.getPitch * 0.8F)
                }
            }
            if (!player.abilities.isCreativeMode) stack.shrink(1)
            ActionResultType.SUCCESS
        }

        if (d < 1 && place(false) == ActionResultType.SUCCESS) return ActionResultType.SUCCESS

        pos = pos.offset(context.getFace)
        place(true)
    }

    /**
     * Create a new part based on the placement information parameters.
     */
    def newPart(item: ItemStack, player: PlayerEntity, world: World, pos: BlockPos, side: Int, vhit: Vector3): TMultiPart
}

File: src/main/java/codechicken/multipart/api/ItemMultipart.java
package codechicken.multipart.api;

import codechicken.multipart.api.part.MultiPart;
import codechicken.multipart.block.TileMultipart;
import codechicken.multipart.util.MultipartPlaceContext;
import net.minecraft.core.BlockPos;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.context.UseOnContext;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.SoundType;
import org.jetbrains.annotations.Nullable;

/**
 * Created by covers1624 on 1/1/21.
 */
public abstract class ItemMultipart extends Item {

    public ItemMultipart(Properties properties) {
        super(properties);
    }

    @Nullable
    public abstract MultiPart newPart(MultipartPlaceContext context);

    @Override
    public InteractionResult useOn(UseOnContext context) {

        MultipartPlaceContext ctx = new MultipartPlaceContext(context);

        if (ctx.getHitDepth() < 1 && place(ctx)) {
            return InteractionResult.SUCCESS;
        }
        return place(ctx.applyOffset()) ? InteractionResult.SUCCESS : InteractionResult.FAIL;
    }

    private boolean place(MultipartPlaceContext context) {
        Level world = context.getLevel();
        BlockPos pos = context.getClickedPos();

        MultiPart part = newPart(context);
        if (part == null || !TileMultipart.canPlacePart(context, part)) return false;

        if (!world.isClientSide) {
            TileMultipart.addPart(world, pos, part);
            SoundType sound = part.getPlacementSound(context);
            if (sound != null) {
                world.playSound(null, pos, sound.getPlaceSound(),
                        SoundSource.BLOCKS, (sound.getVolume() + 1.0F) / 2.0F, sound.getPitch() * 0.8F);
            }
        }
        if (!context.getPlayer().getAbilities().instabuild) {
            context.getItemInHand().shrink(1);
        }
        return true;
    }
}
